clang/lib/Sema/CheckedCAnalysesPrepass.cpp

```diff
@@ -54,13 +62,104 @@ class PrepassHelper : public RecursiveASTVisitor<PrepassHelper> {
       // We only add the V => E pair to the VarUses map if:
       // 1. E is within a declared bounds expression, or:
       // 2. V has a declared bounds expression.
-      if (!InBoundsExpr && !V->hasBoundsExpr())
-        return true;
-      if (Info.VarUses.count(V))
+      if (VarWithBounds || V->hasBoundsExpr()) {
+        if (!Info.VarUses.count(V))
+          Info.VarUses[V] = E;
+      }
+
+      // We add VarWithBounds to the set of all variables in whose bounds
+      // expressions V occurs.
+      if (VarWithBounds) {
+        auto It = Info.BoundsVars.find(V);
+        if (It != Info.BoundsVars.end())
+          It->second.insert(VarWithBounds);
+        else {
+          VarSetTy Vars;
+          Vars.insert(VarWithBounds);
+          Info.BoundsVars[V] = Vars;
+        }
+      }
+
+      return true;
+    }
+
+    bool ProcessWhereClause(WhereClause *WC) {
+      if (!WC)
         return true;
-      Info.VarUses[V] = E;
+
+      for (WhereClauseFact *Fact : WC->getFacts()) {
+        if (BoundsDeclFact *BDF = dyn_cast<BoundsDeclFact>(Fact)) {
+          VarDecl *V = BDF->Var;
+          BoundsExpr *B = BDF->Bounds;
+
+          VisitVarDecl(V);
+          VarWithBounds = V;
+          TraverseStmt(B);
+          VarWithBounds = nullptr;
+        }
+      }
+
       return true;
     }
+
+    bool VisitNullStmt(NullStmt *S) {
+      // Process any where clause attached to a NullStmt. For example,
+      // _Where p : bounds(p, p + 1);
+      return ProcessWhereClause(S->getWhereClause());
+    }
+
+    bool VisitValueStmt(ValueStmt *S) {
+      // Process any where clause attached to a ValueStmt. For example,
+      // x = 1 _Where p : bounds(p, p + 1);
+      return ProcessWhereClause(S->getWhereClause());
+    }
+
+    void DumpBoundsVars(FunctionDecl *FD) {
+      OS << "--------------------------------------\n"
+         << "In function: " << FD->getName() << "\n"
+         << "BoundsVars:\n";
+
+      // Info.BoundsVars is a map of VarDecls (keys) to a set of VarDecls
+      // (values). So there is no defined iteration order for its keys or
+      // values. So we copy the keys to a vector, sort the vector and then
+      // iterate it. While iterating each key we also copy its value (which is
+      // a set of VarDecls) to a vector, sort the vector and iterate it.
+      VarListTy Vars;
+      for (const auto item : Info.BoundsVars)
+        Vars.push_back(item.first);
+
+      SortVars(Vars);
+
+      for (const auto V : Vars) {
+        OS << V->getQualifiedNameAsString() << ": { ";
+
+        VarListTy InnerVars;
+        for (const auto item : Info.BoundsVars[V])
+          InnerVars.push_back(item);
+
+        SortVars(InnerVars);
```

Comment:
Note: If we have `n` variables and each one occurs in the bounds expressions of the other, then this nested call to `SortVars` may have a running time of O(n<sup>2</sup>log<sup>2</sup>n). In practice, this might be a rare scenario.

I am open to ideas to speed this up :)